### ISSUE TITLE
Support gRPC over HTTP/3

### ISSUE THREAD

**Issue by johanbrandhorst** [grpc/grpc#19126]
I'm creating this issue to explore the possibility of developing a spec for gRPC-over-HTTP/3.

HTTP/3 is currently an IETF draft, available to view at https://quicwg.org/base-drafts/draft-ietf-quic-http.html.

**Comment by sanjaypujare** [comment]
Great. I suspect there are others who are interested in this as well. Please share your findings here.

**Comment by DaKellyFella** [comment]
Sounds good. I'm interested and would like to know what your thoughts are on this.

**Comment by AspirinSJL** [comment]
@a11r Please share some thoughts on this.

**Comment by rogchap** [comment]
According to this offical blog post; QUIC is already supported in the other Stacks (C-based and Java) https://grpc.io/blog/grpc-stacks/

Makes sense to have a Go implementation too.

**Comment by pgrosu** [comment]
Maybe take a look at this: https://github.com/gfanton/grpc-quic

**Comment by jtattermusch** [comment]
CC @shirhatti

**Comment by jlisee** [comment]
Can gRPC over HTTP/3 support browser based clients talking to servers without [a proxy layer](https://github.com/grpc/grpc-web#2-run-the-server-and-proxy)?  

The requirement of proxies makes gRPC more difficult than a HTTP or WebSocket based solution for outside the datacenter or embedded projects.  gRPC used only a 1 or 2 features of HTTP/2 that are not supported by browsers.  It's my hope gRPC over HTTP/3 can be created in a 100% browser compatible way.

**Comment by johanbrandhorst** [comment]
> Can gRPC over HTTP/3 support browser based clients talking to servers without [a proxy layer](https://github.com/grpc/grpc-web#2-run-the-server-and-proxy)?

I think this is unlikely to change. The difficulty is in the power of the browser APIs. You can't make a network request and enforce its protocol (HTTP/2 or HTTP/3) from within the browser. Until this changes, pure gRPC from the browser will be impossible.

**Comment by tvvignesh** [comment]
Hi guys. Any ideas on how this can be taken forward? Also, I was wondering - Now that http/3 is solving most of the problems with the transport layer which existed before, will there be any significant performance difference when doing REST over http/3 vs GRPC over http/3 (if and when implemented)?

**Comment by eaglexiang** [comment]
Maybe helpful: [https://github.com/lucas-clemente/quic-go](https://github.com/lucas-clemente/quic-go)

**Comment by shenshouer** [comment]
It's great thing, Any plan for this?

**Comment by vodevil** [comment]
Http3 is UDP + Quic, so a i think grpc will be a legacy tech when http3 will land.

**Comment by hnasr** [comment]
> > Can gRPC over HTTP/3 support browser based clients talking to servers without [a proxy layer](https://github.com/grpc/grpc-web#2-run-the-server-and-proxy)?
> 
> I think this is unlikely to change. The difficulty is in the power of the browser APIs. You can't make a network request and enforce its protocol (HTTP/2 or HTTP/3) from within the browser. Until this changes, pure gRPC from the browser will be impossible.

Why is it impossible? The browser can ask for gRPC protocol as part of the TLS handshake through the ALPN extension, just like they do today with HTTP/2. They can downgrade to a supported protocol if the server doesn’t support gRPC. 

I think the challenge for supporting gRPC on regular web traffic is the lack of schema since gRPC requires well defined schema through protobuff which the web just doesn’t have. 

I might be wrong though.

**Comment by ansarizafar** [comment]
> Http3 is UDP + Quic, so I think grpc will be a legacy tech when http3 will land.

Is this true? I am thinking about using Asp.met core Grpc with Flutter but If the above statement is true then I will go back to Rest apis of graphql.

**Comment by derekperkins** [comment]
@ansarizafar No, that is completely untrue. This issue is about using http/3 with gRPC. gRPC is about defining an api contract - the underlying network transport is separate from that.

**Comment by djc** [comment]
I am working on HTTP/3 support within hyper based on quinn-h3.

**Comment by nikooo777** [comment]
has there been any progress on this by any chances?

**Comment by kellengreen** [comment]
Chrome is currently at origin trial status with `QuicTransport`. Assuming this lands in other browsers, it could perhaps be a way to facilitate gRPC communication over QUIC.

**Comment by sequoiar** [comment]
Any update on support gRPC over HTTP3/QUIC ? the main concern is HTTP2 has known TCP Head-Of-Line blocking issue.

**Comment by libratiger** [comment]
is there any conclusion for  this issue?

**Comment by djc** [comment]
While I made some progress with building hyper support for HTTP/3 on top of Quinn, hyper decided to go a different route, and the h3 effort is now ongoing in https://github.com/hyperium/h3/. Please track that project.

**Comment by larspaulsen** [comment]
QUIC was approved as an IETF standard May 27 2021.

For more information:
https://en.wikipedia.org/wiki/QUIC
https://www.ietf.org/rfc/rfc9000.html
https://datatracker.ietf.org/doc/rfc9000/

I assume this group is aware of the benefits of QUIC, such as:

Seamless underlying connection media switching, such as switching between: Ethernet <-> Wi-Fi <-> Cellular, etc.
Avoid head-of-line-blocking issues.
Quicker reestablishment of known connections.
And many other important features...
Describe the solution you'd like
Support for specifying QUIC (HTTP/3) as the underlying channel transport protocol.

No alternatives, other than to use the less desirable existing support for TCP.

**Comment by djc** [comment]
Just to be fully explicit about this, QUIC and HTTP/3 are two different protocols. QUIC is a new transport protocol (TCP layer, but with parallel streams), HTTP 3 is basically the familiar HTTP semantics reformulated on top of QUIC. Note that while QUIC has standardized in the aforementioned RFCs, HTTP/3 standards have not yet been published.

**Comment by alercah** [comment]
Since gRPC basically uses HTTP/2 as a transport protocol, and QUIC is intended for use as a transport protocol with support for most of the key transport features, does it make more sense to avoid the extra layer and define gRPC-over-QUIC rather than gRPC-over-HTTP/3?

**Comment by kellengreen** [comment]
> does it make more sense to avoid the extra layer and define gRPC-over-QUIC

Yes I'm assuming QUIC should be the desired outcome of this proposal.

**Comment by JamesNK** [comment]
Why should gRPC use QUIC directly?

gRPC uses HTTP features like headers and URLs. If gRPC was built on QUIC directly then it would have to re-invent lots of HTTP features for no benefit. Using HTTP/3's QPACK header compression is much better than gRPC coming up with its own way for sending headers that loses compression for example.

gRPC uses HTTP/2, and wasn't built directly on TCP sockets, to get the benefits of HTTP/2.
gRPC should use HTTP/3, and not build directly on QUIC, to get the benefits of HTTP/3.

**Comment by eaglexiang** [comment]
> Why should gRPC use QUIC directly?
> 
> gRPC uses HTTP features like headers and URLs. If gRPC was built on QUIC directly then it would have to re-invent lots of HTTP features for no benefit. For example, using HTTP/3's QPACK header compression is much better than gRPC coming up with its own way for sending headers that loses compression.
> 
> Think of it like this:
> 
> * gRPC uses HTTP/2, and wasn't built directly on TCP sockets, to get the benefits of HTTP/2.
> * gRPC should use HTTP/3, and not build directly on QUIC, to get the benefits of HTTP/3.

What about `smart congestion control`?

For me, gRPC was used on a unstable TCP layer in the past, whose low performance made me sad. We built a custom UDP layer to solve the performance problem, which is effective but complicated.

**Comment by hnasr** [comment]
I might be missing something, but isn't HTTP/3 technically supported through the Cronet transport? 
https://grpc.io/blog/grpc-stacks/#wrapped-languages
https://grpc.github.io/grpc/core/md_doc_core_transport_explainer.html

gRPC transport is pluggable and you can replace the core HTTP/2 transport implementation with Cronet which supports QUIC and HTTP/3. 

Is this issue specifically to create a new built-in transport for HTTP/3?

**Comment by ejona86** [comment]
gRPC is built on HTTP/1.1 semantics (even if our implementation only supported HTTP/2 and the HTTP/1 ecosystem makes bidi far too difficult, and if the spec says "HTTP/2" more places than it should), so any QUIC usage would be "HTTP/3" not "raw QUIC".

And yes, this has been possible for a long time via Cronet. Mobile was the most interesting use-case for QUIC (this support pre-dates HTTP/3).

I believe there's some doubt whether QUIC-in-the-datacenter provides much benefit. Mobile is already handled with Cronet. So that would mean native QUIC would mostly be for "across the internet" connections between servers.

**Comment by eaglexiang** [comment]
> I might be missing something, but isn't HTTP/3 technically supported through the Cronet transport?
> https://grpc.io/blog/grpc-stacks/#wrapped-languages
> https://grpc.github.io/grpc/core/md_doc_core_transport_explainer.html
> 
> gRPC transport is pluggable and you can replace the core HTTP/2 transport implementation with [Cronet](https://developer.android.com/guide/topics/connectivity/cronet#features) which supports QUIC and HTTP/3.
> 
> Is this issue specifically to create a new built-in transport for HTTP/3?

Cronet is just a Web lib on mobiles, is it comfortable to work as a listener on server? QUIC is not only a Web protocol but also a Transport protocol. Why not take its features like `user space flow-control`?

**Comment by ejona86** [comment]
> Cronet is just a Web lib on mobiles, is it comfortable to work as a listener on server?

Cronet is client-only. When using Cronet on client-side you generally need to have a load balancing proxy between the client and the gRPC server anyway. That proxy would speak QUIC/HTTP/3 and then translate to HTTP/2 when communicating with the gRPC server. Using HTTP/2 from the LB to the backend is fine.

> QUIC is not only a Web protocol but also a Transport protocol. Why not take its features like user space flow-control?

I don't think there's much interest from gRPC devs in having a version of gRPC that uses QUIC without HTTP/3. If you year a gRPC dev say "QUIC" they probably mean "HTTP/3," since we were talking about it before there was any such distinction. User-space flow control is still available within HTTP/3 and existed in HTTP/2. Sticking with HTTP/3 means we maintain compatibility with grpc-over-HTTP/2, it will continue working with generic load balancers, and there's only minor discussion points on how to behave (apparently mostly relating to deadlines/cancellation).

**Comment by larspaulsen** [comment]
_"I don't think there's much interest from gRPC devs in having a version of gRPC that uses QUIC without HTTP/3. If you year a gRPC dev say "QUIC" they probably mean "HTTP/3," since we were talking about it before there was any such distinction. User-space flow control is still available within HTTP/3 and existed in HTTP/2. Sticking with HTTP/3 means we maintain compatibility with grpc-over-HTTP/2, it will continue working with generic load balancers, and there's only minor discussion points on how to behave (apparently mostly relating to deadlines/cancellation)."_

I completely agree with Eric (ejona86), lets not try to complicate things unnecessarily. The most resource effective, least risky and logic path is to 'simply' add configuration options for when instantiating gRPC channels to either use either a) HTTP/2 or b) HTTP/3, when subsequently opening the channel. The HTTP/3 layer will then know how to most effectively manage the use of the underlying QUIC based virtual streams.

The proxy in the cloud will typically convert to legacy protocols as needed as illustrated below:
![HighLevelQuicEcoSystem](https://user-images.githubusercontent.com/13814018/126850106-2f071267-6af9-4edf-9b58-c2a081461546.png)

**Comment by JamesNK** [comment]
FYI I've started a proposal for gRPC over HTTP/3: https://github.com/grpc/proposal/pull/256

Please keep any comments there constructive 😄

Thumbs up reactions = good
Commenting that you want this or "+1" = noise

**Comment by AmSoMad** [comment]
Is this issue still going on?

**Comment by zs-dima** [comment]
+1

**Comment by libratiger** [comment]
> 
I just wonder what is the main tech challenge to for this feature in the core gRPC?

**Comment by larspaulsen** [comment]
I see Cronet already supports QUIC but I am not familiar with what else perhaps is missing.

@ejona86 Are you or would you be able to point to someone who can provide an overview of the architectural changes needed for Core gRPC to support HTTP/3. This is so all the deriving language implementations can provide wrappers for the core implementation and do not need to create their own implementation, like .NET did (I know they did it because they use a different underlying stack if I understand things correctly)?

**Comment by ejona86** [comment]
> Are you or would you be able to point to someone who can provide an overview of the architectural changes needed for Core gRPC to support HTTP/3.

Are you actively interested in contributing the work? Most of the problem is actually doing the work and maintaining it. I don't think there's much necessary from an architectural perspective, at least at this point.

> The rest of the languages implementation are in reality wrappers on top of the underlying Core C++ based implementation

Not true for Java, Go, .Net, Node.JS. The wrapper languages probably benefit little from QUIC because performance gains are probably in the noise: Python, Ruby, PHP. Obj C could benefit, but it can use Cronet today and there are serious deployment concerns it would face with its own QUIC stack.

**Comment by larspaulsen** [comment]
Eric, is it possible you could accept the invitation I send you a while back through LinkedIn, so we can take a short out-of-band discussion of what you envision this work would consist of?

----------------

With reference to the following, now supposedly outdated, architectural diagram:

https://grpc.io/blog/grpc-stacks/

then if I understand your answer correctly then when Cronet added support for HTTP/3 it got added on top of the shown QUIC component box, and that the .NET team used that in their implementation described below:

https://github.com/JamesNK/proposal/blob/jamesnk/http3/G2-http3-protocol.md

If that is correctly understood then according to the above referenced diagram everything needed in the 'gRPC C Surface and C/C++ Core' layer should already been implemented and tested by the .NET team. Therefore, the two next ‘easiest’ commonly used languages to add support for HTTP/3 would be the a) C++ and b) Java implementations. Correct?

Quick question: What is it the current languages having Cronet shown underneath them uses Cronet/TCP for? Or is the drawing missing a TLS and HTTP/2 box inside the Cronet box? (Supposedly used for example for the Android based usage).

**Comment by Jeffwan** [comment]
Could someone help me understand the benefits to use gRPC over QUIC in micro-service communication? I think the above communication talks a lot from mobile to micro-service side. I am pretty sure QUIC will help in this case. But for micro-services, it may be different. Most of the applications are deployed in one data center and the network is pretty stable for service internal communication. In that case, I just need QUIC from mobile to gateway?  What's the real benefits for micro service side internal communication (let's say service are connected via service mesh)? 

![image](https://user-images.githubusercontent.com/4739316/184698248-a264b378-df51-4a42-a7f0-951aea65f006.png)

**Comment by dfinkel** [comment]
Yes, there are several advantages.

The first that comes to mind is that gRPC over QUIC would eliminate head of line blocking issues associated with HTTP/2's use of a single TCP socket per-channel. (in theory this could eliminate the need for some connection-pools between services)

QUIC's userspace handling of congestion control allows for finer control of prioritization in when the available bandwidth becomes constrained.

For services that don't use LBs and talk directly to backend instances the use of UDP can also allow for multiplexing with filedescriptors, thus eliminating a bottleneck. (particularly in cases where ulimits default to insanely low values)

**Comment by kellengreen** [comment]
Additionally, gRPC over QUIC could open the door for browsers to join in on the fun via the `QuicTransport` proposal.

**Comment by larspaulsen** [comment]
Other than the fact that is is currently only supported for C#, then there are in my opinion no good reason not to use QUIC for inter-service communication, which is what for example Meta (Facebook) has chosen to do for virtually all their services.

**Comment by eaglexiang** [comment]
I have a plan to deploy gRPC on QUIC with Go.

`grpc-go` supports DialerOptions with client option [WithDialer](https://github.com/net-bridge/netbridge-client/blob/2d124647564c36007d840db6a83855c6ba79dbc1/vendor/google.golang.org/grpc/dialoptions.go#L380).

```go
func WithDialer(f func(string, time.Duration) (net.Conn, error)) DialOption {
```

`grpc-go` supports custom Listener with server Method [Serve](https://github.com/net-bridge/netbridge-server/blob/7b0f2c8553f63b3ad8480254c3a79d76f6316eef/vendor/google.golang.org/grpc/server.go#L747)

```go
func (s *Server) Serve(lis net.Listener) error {
```

`quic-go` supports [integration without HTTP3](https://github.com/lucas-clemente/quic-go/blob/master/example/echo/echo.go). 

My simple plan:

1. build Dialer and Listener with QUIC without HTTP3.
2. set Dialer and Listener with gRPC options.

It works.

**Comment by eaglexiang** [comment]
> I have a plan to deploy gRPC on QUIC with Go.
> 
> `grpc-go` supports DialerOptions with client option [WithDialer](https://github.com/net-bridge/netbridge-client/blob/2d124647564c36007d840db6a83855c6ba79dbc1/vendor/google.golang.org/grpc/dialoptions.go#L380).
> 
> ```go
> func WithDialer(f func(string, time.Duration) (net.Conn, error)) DialOption {
> ```
> 
> `grpc-go` supports custom Listener with server Method [Serve](https://github.com/net-bridge/netbridge-server/blob/7b0f2c8553f63b3ad8480254c3a79d76f6316eef/vendor/google.golang.org/grpc/server.go#L747)
> 
> ```go
> func (s *Server) Serve(lis net.Listener) error {
> ```
> 
> `quic-go` supports [integration without HTTP3](https://github.com/lucas-clemente/quic-go/blob/master/example/echo/echo.go).
> 
> My simple plan:
> 
> 1. build Dialer and Listener with QUIC without HTTP3.
> 2. set Dialer and Listener with gRPC options.
> 
> It works.

There is my usecase:

* [ServerSide](https://github.com/net-bridge/netbridge-server/blob/7b0f2c8553f63b3ad8480254c3a79d76f6316eef/grpc/grpc.go#L83)
* [ClientSide](https://github.com/net-bridge/netbridge-client/blob/2d124647564c36007d840db6a83855c6ba79dbc1/grpc/init.go#L36)

**Comment by larspaulsen** [comment]
Hi

**Comment by TBBle** [comment]
If I'm reading grpc-go's `Serve` code correctly, that example is demonstrating HTTP/2 over QUIC. See `func (s *Server) handleRawConn` which calls `func (s *Server) newHTTP2Transport`.

So this seems like the worst-of-both-worlds, since HTTP/2 won't be able to take advantage of QUIC in any way.

**Comment by eaglexiang** [comment]
Maybe someone just need:

UDP transmission, usermode congestion control with QUIC.

Simple RPC development with gRPC.

This couple did help me save much time from coding :)

**Comment by AmSoMad** [comment]
My guess is that next year September 2023 there will be a new "QUICG" standard that combines UDP and Grpc.
So this problem will be solved. And my comment will be a post foretelling a great future.

We can use HTTP2 & 3!! this QUICG !

1. Enable HTTP3 by default in all browsers
2. Enable HTTP3 for all server developers live service
3. blah blah blah...

**Comment by TBBle** [comment]
We have a [gRPC over HTTP/3 proposal](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md) and [example implementation](https://github.com/grpc/grpc-dotnet) already, 2 years before September 2023.

Is there a reason not to start prototyping this proposal in other language stacks that have HTTP/3 support, compared to reimplementing HTTP/3-level features in a direct gRPC-QUIC binding?

**Comment by larspaulsen** [comment]
TBBle, I agree lets not try to reinvent the already proven wheel.

**Comment by tony-clarke-amdocs** [comment]
@eaglexiang Your github repositories are private, so cannot see your [ServerSide](https://github.com/net-bridge/netbridge-server/blob/7b0f2c8553f63b3ad8480254c3a79d76f6316eef/grpc/grpc.go#L83)
[ClientSide](https://github.com/net-bridge/netbridge-client/blob/2d124647564c36007d840db6a83855c6ba79dbc1/grpc/init.go#L36). Can you share something publicly?

**Comment by cubxxw** [comment]
Is this question（issue 19126) conclusive？

**Comment by zs-dima** [comment]
@cubxxw 
Sure, we are waiting when it will be implemented.

**Comment by cubxxw** [comment]
> @cubxxw Sure, we are waiting when it will be implemented.

Okey,how long and in what version?

**Comment by hwkim1127** [comment]
Is there any progress?

**Comment by mylaluna** [comment]
Any news?

**Comment by MadDeveloper** [comment]
Any update?

**Comment by masx200** [comment]
Any update?

**Comment by begmaroman** [comment]
Any update?

**Comment by TBBle** [comment]
Looks like no update, all the current GRPC repositories except "gRPC for .NET" are waiting for someone who has the time to implement the existing HTTP/3 proposal (and possibly also waiting for underlying HTTP/3 support in their language), i.e. no apparent change from https://github.com/grpc/grpc/issues/19126#issuecomment-1257383670.

**Comment by ejona86** [comment]
@TBBle, don't forget about https://github.com/grpc/grpc/issues/19126#issuecomment-884541156 . Languages have had HTTP/3 support on mobile for a very long time. I've added notes to the top comment to hopefully avoid some confusion.

As of yet, HTTP/3 in the datacenter doesn't seem to provide much value gain, especially relative to the effort in code, rollout, and maintenance. The discussion here hasn't characterized any serious need for HTTP/3 in the datacenter.

I'm tempted to close this issue. A new issue could be much more focused for someone to make a case why HTTP/2 is poorly serving them in the data center. (E.g., head of line blocking only matters if there are dropped packets; to predict that HTTP/3 would be of benefit, you'd probably want to mention drop rates.)

**Comment by robindegen** [comment]
I can indeed confirm @AttilaTheFun 's statement. This is exactly the usecase I have in mind. I would love to have to stop fiddling around with rest api's and just make a grpc connection from a mobile app to a backend. It would make so many things so much easier.

**Comment by ejona86** [comment]
@AttilaTheFun, why do native clients need HTTP/3 to communicate? Native clients already support HTTP/2. Are you saying, "there is a rise of gRPC communication that happens across the open Internet, unrelated to mobile, and this communication can experience poor networks"? Can you explain which types of networks you're concerned about? How is the movement away from REST involved? Is it "I used HTTP/3 with my HTTP client, but I lose that with gRPC"?

@robindegen, you said "mobile," where HTTP/3 is already supported in multiple languages. Are you wanting it in a language where Cronet isn't supported? How is REST related to HTTP/3? Mentioning REST makes it seem you can't use gRPC at all on the client.

Since REST/GraphQL was mentioned, which is heavily used on browsers, I'll note that "browser support without a very specific proxy" would be https://github.com/grpc/grpc/issues/29818 .

**Comment by zs-dima** [comment]
@AttilaTheFun @robindegen +1
gRPC is very useful for frontend to backend communication.

**Comment by AttilaTheFun** [comment]
@ejona86 my understanding is that HTTP/3 (and QUIC) is especially impactful for communication from clients with unreliable or frequently changing network conditions.

This is extremely common on mobile devices which are very frequently changing networks and network conditions. E.g. home WiFi to cellular to poor cellular conditions on the subway to work wifi.

I work at Snapchat and we use HTTP/3 for loading images / videos from our servers typically via regular HTTP requests, and we use GRPC for the majority of our client / server communication. It would be great if we could use HTTP/3 for that GRPC communication as well.

**Comment by ejona86** [comment]
@AttilaTheFun, gRPC has supported HTTP/3 for mobile since 2017 or so, using Cronet. Seems you are likely already using Cronet for HTTP, so this is a great approach as it can share connections/state between regular HTTP and gRPC. What language were you needing HTTP/3 support on mobile?

**Comment by AttilaTheFun** [comment]
@ejona86 for Go on the backend:

https://github.com/grpc/grpc-go/issues/5186

The grpc-go ticket was closed in favor of this one.

Also the iOS native networking stack supports HTTP/3 but grpc-swift does not AFAIK:

https://github.com/grpc/grpc-swift

Also Cronet is a *very* heavy dependency and if it was possible to just use the platform support for HTTP/3 when available that would be preferred.

**Comment by ItalyPaleAle** [comment]
Our team at Microsoft is working on an Azure service whose data plane operations are exposed using gRPC. The service is meant to be consumed by other applications that are running in workstations and servers running in heterogeneous environments (Kubernetes, VMs, serverless, Raspberry PIs in a closet, etc) and in various geographical location. Though not a strict requirement, HTTP/3 support for gRPC would be welcome in our case too, and that would require implementation in gRPC for Go, Java, Python, etc.

**Comment by StevenACoffman** [comment]
I am working on embedded devices using gRPC, and Cronet is too heavy for our purposes. (grpc-go)

**Comment by badarsebard** [comment]
Similar to the embedded device use case mentioned above, there are use cases for many remote service vendors like MSSPs who deploy their technology to client environments which are not under their control, and often geographically dispursed but utilizing an edge network isn't viable from a cost standpoint. Use of gRPC has significant advantages in the control of the remote devices, simplifies their development, and transport over QUIC alleviates many of the networking issues and eliminates the need for complicated and brittle tunneling methods.

**Comment by ejona86** [comment]
(Thank you everyone that is including languages of interest in your comments. That is helpful)

To summarize a common thread, there are IoT/embedded/workstation use cases that would benefit from HTTP/3. In my opinion they probably aren't _as bad_ as mobile, with probably less focus on transitioning between networks, but still can experience crummy internet connections and thus packet loss. And this is spanning C/Java/Go (C includes Python), for likely the typical reasons we support so many languages.

One important part of the mobile use-case is there's no need for client-side load balancing. That makes it much easier to use an existing library. It sounds like that is likely true for non-datacenter cases in general.

------

> eliminates the need for complicated and brittle tunneling methods.

@badarsebard, could you explain how HTTP/3 helps with tunneling?

> @ejona86 for Go on the backend:

@AttilaTheFun, since almost no mobile use-case that is interested in QUIC want to do client-side load balancing, I think we'd expect you to use a reverse proxy in front of your backend. And that seems appropriate for any HTTP/3-across-the-internet use-case. The proxy can use HTTP/2 to the backend. (Even if not ideal, it seems likely acceptable.) So grpc-go HTTP/3 on server-side seems low frequency use-case. Although maybe it could be working already if you used grpc-go's Server.ServeHTTP with a go-quic server.

> Also Cronet is a very heavy dependency and if it was possible to just use the platform support for HTTP/3 when available that would be preferred.

That conversation is a very deep iOS discussion, and probably best not done here since it is mobile-related. Using the platform support would be much easier for gRPC, but I expect there are limitations in what it can do (and I expect grpc-swift folks are already very well aware). HTTP/3 implementations are heavy-weight in general, because of the TLS integration required. We wouldn't be able to do much better size-wise.

> Kubernetes, VMs, serverless

@ItalyPaleAle, I agree those could be cross-cloud or on-prem, but I'd expect them to generally have pretty reasonable internet. But, yeah, with "workstations" and "Raspberry PIs in a closet" then things start skewing toward worse connections.

@badarsebard seems similar, skewed toward more embedded.

> I am working on embedded devices using gRPC, and Cronet is too heavy for our purposes. (grpc-go)

@StevenACoffman, we don't have cronet support in grpc-go. grpc-go could be the one language that doesn't see as big of a binary size increase with HTTP/3 support, because the TLS library could be reused. There might be options here to reuse quic-go, but there's some architectural questions and it _might_ help to have https://github.com/grpc/proposal/pull/103 .

**Comment by badarsebard** [comment]
> @badarsebard, could you explain how HTTP/3 helps with tunneling?

First, I will acknowledge that tunneling is possible with Hhttp/2 and current gRPC implementations. However, http/3 introduces two advantages when doing so. First, there are some security issues from http/2 that are solved. Second, using http/2 can be challenging in practice because many networks will either terminate long-lived http/2 connections or heavily restrict the outbound destinations. Use of UDP for a transport protocol avoids both of those issues and makes establishing and managing a tunnel much simpler.

**Comment by alippai** [comment]
For the datacenter use-case: RoCE v2 uses UDP similarly to HTTP3. Did some see a benchmark for http2 vs http3 in a datacenter (100-800g)? We know it doesn’t help webpage load time, but it doesn’t mean it won’t help other workloads.

**Comment by isheff** [comment]
It is often useful to send RPCs to large numbers of servers (including in P2P systems), and maintaining open TCP connections becomes significant overhead. QUIC (or, hopefully, HTTP/3) can help a lot here, and I'd like to be able to use gRPC natively (without having to pretend to be an Android device with Cronet).

**Comment by TBBle** [comment]
> @TBBle, don't forget about [#19126 (comment)](https://github.com/grpc/grpc/issues/19126#issuecomment-884541156) . Languages have had HTTP/3 support on mobile for a very long time. I've added notes to the top comment to hopefully avoid some confusion.

My main GRPC development exposure has been in either Go or Python GRPC servers so I'm not familiar with Cronet and how exactly it is integrated in the C library, but if we already have at least a client-side GRPC/HTTP-3 implementation by using Cronet for the C-library-wrapping implementations, is it interoperable with the .NET GRPC/HTTP3 implementation and [G2: gRPC over HTTP/3](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md), or do you always require a gateway to unwrap from HTTP/3 and rewrap in HTTP/2? The linked discussion about Cronet predated the proposal, and wasn't mentioned in it.

My main _future_ interests are Go (which does not use the C wrapper) and Rust (which is not part of this repo), primarily Windows client and either Windows or Linux Servers. Both languages have WIP http3 support but neither appears to be at the point where [G2: gRPC over HTTP/3](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md) can be implemented in their main GRPC implementations; Go also has quic-go, but I thought I saw somewhere that grpc-go weren't keen on relying on that library, and would prefer the coming native http3 support. However, I can't immediately find that comment, so maybe it was a dream. (Also, as noted above, the quic-go request was punted back here, but I'm not clear why that was done if it's not relying on wrapping the C library. Was there some unexpressed doubt about [G2: gRPC over HTTP/3](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md)?)

**Comment by ejona86** [comment]
> RoCE v2 uses UDP similarly to HTTP3

@alippai, RDMA results are generally not directly applicable to RPC. And it seems it isn't intended to go between data centers (disallowed packet reordering). I expect it'd be very difficult to draw conclusions relevant to gRPC.

> if we already have at least a client-side GRPC/HTTP-3 implementation by using Cronet for the C-library-wrapping implementations, is it interoperable with the .NET GRPC/HTTP3 implementation and G2: gRPC over HTTP/3

@TBBle, I don't know if anyone has tried to communicate between them, but they should be compatible. gRPC really is using HTTP/1.1 semantics, so HTTP/3 was pretty well defined for gRPC as soon as it existed. There are some small things you can decide how to handle, since HTTP/2 and 3 have more expressive error codes, but overall if you just think "how would this be converted from HTTP/1.1?" you'd get the right idea.

**Comment by TBBle** [comment]
I'm surprised GRPC doesn't have an interoperability test suite by now. I guess it's because so many implementations share the C core library, it hasn't been a priority.

**Comment by alippai** [comment]
@ejona86 I'm not sure what your message means. Can you share why do you claim HTTP/3 in datacenter is not relevant?

**Comment by ejona86** [comment]
@TBBle, we have interop tests. We actually test across all languages and versions. We are still testing against 1.0. But it doesn't include quic from mobile to .net quic server. It focuses on HTTP/2.

@alippai, moving IPs and packet loss are significantly more limited problems within a data center than mobile→datacenter communication. If you think something should be better, of the size and complexity as this, you need to be able to argue why it should help, and quantifiable data would go a long way. I'm aware of some experimentation done with quic in a datacenter and the results were lackluster. Not bad, but "not obviously so much better that it is immediately worth lots of investment" sort of idea. So if you know how quic is relevant for you, mention it. (And I'm not ignoring the reduced fd usage one person mentioned.)

**Comment by mateli** [comment]
Would it be possible to rely on a VPN to solve the mobile reliability issue? Packet loss may not be a problem within a datacenter but between datacenters is may be as well as all communication with devices. "Mobile" in this context may also include laptops that are using mobile connections that may be unstable. The main advantage of quic is that it is application level where using VPN may not be possible simply because the user lacks admin rights.

**Comment by martin-sucha** [comment]
> As for the category or packages or RPC calls where failure is better than resending packages (for example game network libraries and RPC layers usually has this option) this is not possible with HTTP 2 and probably not HTTP 3 either. To achieve that another message transport layer would be needed.

There is a proposed standard [RFC 9221 An Unreliable Datagram Extension to QUIC](https://www.rfc-editor.org/rfc/rfc9221) and [RFC 9297 HTTP Datagrams and the Capsule Protocol](https://datatracker.ietf.org/doc/rfc9297/)

However, it is probably not useful for gRPC, which needs reliable transport.

**Comment by mateli** [comment]
> > As for the category or packages or RPC calls where failure is better than resending packages (for example game network libraries and RPC layers usually has this option) this is not possible with HTTP 2 and probably not HTTP 3 either. To achieve that another message transport layer would be needed.
> 
> There is a proposed standard [RFC 9221 An Unreliable Datagram Extension to QUIC](https://www.rfc-editor.org/rfc/rfc9221) and [RFC 9297 HTTP Datagrams and the Capsule Protocol](https://datatracker.ietf.org/doc/rfc9297/)
> 
> However, it is probably not useful for gRPC, which needs reliable transport.

That looks like it could open up for low latency communication/RPC using QUIC/HTTP3. The correct way to express this is that it is not useful for gRPC unless gRPC adds unreliable calls to the stack like for example the RPC layer in godot has. Although with QUIC/HTTP3 it should be possible to run gRPC side by side with other communication so that gRPC can be used for calls that need reliability and another delivery mechanism can be used for transferring data where no reliability is preferred.

**Comment by pedrolamarao** [comment]
gRPC over QUIC would benefit from transport-level streaming with no head-of-line blocking.

**Comment by TBBle** [comment]
There's a bunch of discussion on gRPC unreliable transport use-cases in #9493, but it all predates RFC 9221/RFC 9297 and ended with "gRFC needed", which is probably fair.

I'll also mention that I came across [this article published in July](https://kmcd.dev/posts/grpc-over-http3/) which explores existing gRPC/HTTP-3 implementations, including adding HTTP/3 support to [`buf curl`](https://buf.build/docs/reference/cli/buf/curl#http3) which will be handy for validating server-side implementations/deployments.

**Comment by apm1467** [comment]
I can provide a point for using HTTP/3 inside data centers. 

My team at TikTok uses gRPC (Java) for inter-service communication within data centers. The network conditions are usually pretty good, but sometimes the packet drop rate can raise transiently due to incidents like a network device failure. Our experience is that the gRPC traffic throughput can decrease significantly even if the packet drop rate only raises a little. We've observed 30%+ throughput loss when the packet drop rate raised from 0.01% to 0.5%. 

This is presumably due to TCP Head of Line blocking. By switching to HTTP/3, gRPC could be more resilient to such network incidents inside data centers.

**Comment by mateli** [comment]
@apm1467 That's exactly the kind of problem that QUIC were designed to prevent. Although there are some modernization to TCP that may mitigate such problems, such solutions must be supported by the kernels involved. QUIC resides on the application layer and is easy to deploy.

**Comment by LPardue** [comment]
> I can provide a point for using HTTP/3 inside data centers.
> 
> My team at TikTok uses gRPC (Java) for inter-service communication within data centers. The network conditions are usually pretty good, but sometimes the packet drop rate can raise transiently due to incidents like a network device failure. Our experience is that the gRPC traffic throughput can decrease significantly even if the packet drop rate only raises a little. We've observed 30%+ throughput loss when the packet drop rate raised from 0.01% to 0.5%.
> 
> This is presumably due to TCP Head of Line blocking. By switching to HTTP/3, gRPC could be more resilient to such network incidents inside data centers.

I tend to disagree with this hypothesis, especially in the absence of data.

If the network is experiencing congestion, the transport protocols congestion controller will kick in and reduce the sending rate. That's a very likely a cause of a reduction in goodput. 

Head of line blocking avoidance only matters when there are independent multiplexed requests in flight. 

Understanding the concrete benefits for a gRPC use case depends on the measurement methodology. I'd expect that to focus on gRPC latency, such as time to response, and time to completion. Measuring connection throughput is probably not a good indicator for whether HoL avoidance would help.

I think this issue is getting distracted slightly by focusing on lower layer matters. The high order bit is this: its possible today to use a single HTTP/2 connection to multiplex HTTP and gRPC traffic. It should be possible to do the same for HTTP/3, using a formal spec. Without that, application either need to use two connections when one would suffice, or implement the proposal for gRPC over H3, which has all the risks associated with supporting a quasi-spec.

Defining a gRPC over QUIC application protocol does not sound useful to me. You won't get the pooling mentioned above. And you'll have to reinvent all the things that HTTP provides for free. And you'll have to convince vendors to support a non-HTTP ALPN, which opens a whole can of Worms. 

gRPC over HTTP/3 is the obvious path forward.

**Comment by LPardue** [comment]
For all intents and purposes, TCP and QUIC implement the same congestion control mechanisms. If a chosen congestion control algorithm is configured for both, and the tuning parameters are the same, you should expect the congestion response and it's affect on bandwidth utilisation to be the same.

gRPC depends on HTTP semantics. Those can transfer from HTTP/2 to HTTP/3 easily. As demonstrated by the protocol proposal and litany of implementations. It's basically a no brainer.

I don't see the point in litigating about other possible RPC choices on this issue. That work can go happen elsewhere if people want to.

**Comment by LPardue** [comment]
FWIW that paper has several issues in its methodology. I suggest reading any of its claims with care and attention.

As @hnasr rightly highlights, the benefits and the constraints of QUIC are workload dependent. The paper picks weird workloads that are not representative.

I think this thread is overindexing on the binding of HTTP/3 to lower layers in certain kinds of deployments. There's a lot of opportunity for performance tuning, using a slew of engineering solutions. The kernel option should be assumed to offer magic benefits. There's a lot that is being done outside kernel. Microsoft has published some good work, they are not alone https://techcommunity.microsoft.com/t5/networking-blog/making-msquic-blazing-fast/ba-p/2268963

Furthermore, the IETF continues to identify protocol opportunities. Such as substituting QUIC packetization and datagrams, with secure streams. Such an approach can be more suited to a DC. See https://datatracker.ietf.org/doc/html/draft-kazuho-quic-quic-on-streams.

The missing piece is a standard protocol definition for gRPC over HTTP/3. That's the abstraction layer that opens up all these other opportunities underneath.

**Comment by ejona86** [comment]
[gRPC over HTTP/3 is defined](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md). It's basically, "do what you would expect given the HTTP semantics or HTTP/2 behaviors," although with another error code mapping. A proxy with no knowledge of gRPC can convert gRPC between HTTP/2 and HTTP/3 without any real concern.

**Comment by LPardue** [comment]
@ejona86 appreciate your stewardship on the matter. My question is more about how we can move the [proposal](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md) beyond proposal. This might reflect my unfamiliarty with the process more than anything, I read up [here](https://github.com/grpc/proposal?tab=readme-ov-file#process) and think we're at step 6?

The primary difference is in error handling, both the codes and the way stream lifetime is handled. I'm not sure what expectations there are the error codes are handled rigidly as defined in the proposal. 

For instance, gRPC PERMISSION_DENIED is mapped to HTTP/2 INADEQUATE_SECURITY. HTTP/3 has no equivalent error code and the [mapping table](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md#errors) leaves it as n/a. Imagine a proxy that has HTTP/3 on one side and HTTP/2 , if it receives INADEQUATE_SECURITY, what does it do? Can the h3 gRPC app deal with receiving a "random" error code?

**Comment by TBBle** [comment]
> My question is more about how we can move the [proposal](https://github.com/grpc/proposal/blob/master/G2-http3-protocol.md) beyond proposal. This might reflect my unfamiliarty with the process more than anything, I read up [here](https://github.com/grpc/proposal?tab=readme-ov-file#process) and think we're at step 6?

It's in the `Implemented` state, so it's completed the proposal process (see the end of the proposal process page), and it's just up to individual implementation maintainers to implement it, as grpc-dotnet has.

I suspect that's generally blocked on HTTP/3 implementations in the various languages. At least that seems to be true for Go and Rust, the ones I'm personally tracking, which also happen to live outside this repo... (This [blog post series](https://kmcd.dev/series/grpc-over-http%23473/) demonstrates a working gRPC/HTTP3 client _and server_ in Go using a third-party gRPC implementation, for example.)

For language implementations in this repo that are built around on the core C library, then we have _client_ support via Cronet but I think it's only mobile-related targets that package it for users, e.g., Java (for Android), Objective-C, and I noticed Flutter offers it as an experimental feature too.

For general HTTP/3 support on the implementations in this repo, I guess someone could add http/3 support to chttp2 or contribute a similarly in-tree HTTP/3 implementation? I don't know if there's an existing plan here, I haven't noticed one in the issue tracker. I suspect no one currently has both the necessary skills and the need to do the implementation work, or it'd have been done by now.

**Comment by pjebs** [comment]
... So is grpc supported with http3 now in Go using https://github.com/quic-go/quic-go?

I didn't fully understand the discussions above.
